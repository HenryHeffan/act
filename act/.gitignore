html
latex
libact_*.a
i386_*
*.o
act_parse*
act.cy
act_walk_X*
expr.cc
expr.h
act_parse_id.h
Makefile.deps
NOTES
TAGS
*~
Changelog
