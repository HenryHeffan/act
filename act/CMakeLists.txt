add_subdirectory(common)
add_subdirectory(pgen)
add_subdirectory(miniscm)
add_subdirectory(lang)
add_subdirectory(passes)
add_subdirectory(transform)
add_subdirectory(simulation)
add_subdirectory(tech)
add_subdirectory(scripts)
