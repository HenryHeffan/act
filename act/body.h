/*************************************************************************
 *
 *  This file is part of the ACT library
 *
 *  Copyright (c) 2018-2019 Rajit Manohar
 *
 *  This program is free software; you can redistribute it and/or
 *  modify it under the terms of the GNU General Public License
 *  as published by the Free Software Foundation; either version 2
 *  of the License, or (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 51 Franklin Street, Fifth Floor,
 *  Boston, MA  02110-1301, USA.
 *
 **************************************************************************
 */
#ifndef __ACT_BODY_H__
#define __ACT_BODY_H__

#include <act/act_id.h>

/*------------------------------------------------------------------------
 *
 *
 * Act body types
 *
 *
 *------------------------------------------------------------------------
 */

class ActBody {
 public:
  ActBody ();
  virtual ~ActBody();
  
  void Append (ActBody *b);
  ActBody *Tail ();
  ActBody *Next () { return next; }

  virtual ActBody *Clone() { return NULL; }

  virtual void Expand (ActNamespace *, Scope *) { fatal_error ("Need to define Expand() method!"); }

  void Expandlist (ActNamespace *, Scope *);

  virtual void Print (FILE *fp) { }

  void updateInstType (list_t *namelist, InstType *it);

 private:
  ActBody *next;
};


/**
 * Instance 
 */
class ActBody_Inst : public ActBody {
 public:
  ActBody_Inst(InstType *, const char *);
  
  void Expand (ActNamespace *, Scope *);
  Type *BaseType ();
  void Print (FILE *fp);

  InstType *getType () { return t; }
  const char *getName() { return id; }
  void updateInstType (InstType *u);

  ActBody *Clone ();

 private:
  InstType *t;
  const char *id;
};

class ActBody_Attribute : public ActBody {
public:
  ActBody_Attribute(const char *_inst, act_attr *_a, Array *_arr = NULL) {
    inst = _inst; a = _a; arr = _arr;
  }
  void Expand (ActNamespace *, Scope *);
  //void Print (FILE *fp);

  ActBody *Clone ();

private:
  const char *inst;
  act_attr *a;
  Array *arr;
};


class ActBody_Conn : public ActBody {
 public:
  ActBody_Conn(ActId *id1, AExpr *ae) {
    type = 0;
    u.basic.lhs = id1;
    u.basic.rhs = ae;
  }

  ActBody_Conn(AExpr *id1, AExpr *id2) {
    type = 1;
    u.general.lhs = id1;
    u.general.rhs = id2;
  }

  void Print (FILE *fp);
  void Expand (ActNamespace *, Scope *);

  ActBody *Clone();
  
 private:
  union {
    struct {
      ActId *lhs;
      AExpr *rhs;
    } basic;
    struct {
      AExpr *lhs, *rhs;
    } general;
  } u;
  unsigned int type:1; /**< 0 = basic, 1 = general */
};

class ActBody_Loop : public ActBody {
 public:

  enum type {
    SEMI,
    COMMA,			// unused
    AND,			// unused
    OR,				// unused
    BAR				// unused
  };
    
  ActBody_Loop (ActBody_Loop::type _t, 
		const char *_id, 
		Expr *_lo, /* NULL if this is a 0..hi-1 loop */
		Expr *_hi,
		ActBody *_b) {
    t = _t;
    id = _id;
    lo = _lo;
    hi = _hi;
    b = _b;
  }

  void Expand (ActNamespace *, Scope *);

  void Print (FILE *fp);

  ActBody *Clone();

  ActBody *getBody() { return b; }

 private:
  ActBody_Loop::type t;			/**< type of loop */
  const char *id;			/**< loop variable */
  Expr *lo, *hi;		/**< range */
  ActBody *b;			/**< body of the loop */
};

class ActBody_Select_gc {
 public:
  ActBody_Select_gc (const char *_id, Expr *_lo, Expr *_hi,
		     Expr *_g, ActBody *_s) {
    id = _id;
    lo = _lo;
    hi = _hi;
    g = _g;
    s = _s;
    next = NULL;
  }
  ActBody_Select_gc (Expr *_g, ActBody *_s) {
    id = NULL;
    lo = NULL;
    hi = NULL;
    g = _g;
    s = _s;
    next = NULL;
  }
  void Append (ActBody_Select_gc *s) {
    next = s;
  }
#if 0
  ~ActBody_Select_gc () {
    if (next) {
      delete next;
    }
    delete s;
  }
#endif
  int isElse() { return g == NULL ? 1 : 0; }

  ActBody *getBody() { return s; }
  ActBody_Select_gc *getNext() { return next; }

  ActBody_Select_gc *Clone ();

private:
  const char *id;
  Expr *lo, *hi;
  Expr *g;			/**< guard */
  ActBody *s;			/**< statement */
  ActBody_Select_gc *next;		/**< rest of the selection */

  friend class ActBody_Select;
  friend class ActBody_Genloop;
};

class ActBody_Select : public ActBody {
 public:
  ActBody_Select (ActBody_Select_gc *_gc) {
    gc = _gc;
  }
#if 0
  ~ActBody_Select () {
    if (gc) {
      delete gc;
    }
  }
#endif

  void Expand (ActNamespace *, Scope *);

  ActBody *Clone();

  ActBody_Select_gc *getGC() { return gc; }

private:
  ActBody_Select_gc *gc;
};

class ActBody_Genloop : public ActBody {
 public:
  ActBody_Genloop (ActBody_Select_gc *_gc) {
    gc = _gc;
  }
  void Expand (ActNamespace *, Scope *);
  ActBody *Clone();

  ActBody_Select_gc *getGC() { return gc; }

private:
  ActBody_Select_gc *gc;
};

class ActBody_Assertion : public ActBody {
public:
  ActBody_Assertion (Expr *_e, const char *_msg = NULL) {
    type = 0;
    u.t0.e = _e;
    u.t0.msg = _msg;
  }
  ActBody_Assertion (ActId *_id1, ActId *_id2, int op,
		     const char *_msg = NULL) {
    type = 2;
    u.t2.id1 = _id1;
    u.t2.id2 = _id2;
    u.t2.msg = _msg;
    u.t2.op = op;
  }
  ActBody_Assertion (InstType *nu, InstType *old) {
    type = 1;
    u.t1.nu = nu;
    u.t1.old = old;
  }
  ~ActBody_Assertion () {
    if (type == 0) {
      if (u.t0.e) {
	expr_free (u.t0.e);
      }
    }
  }
  void Expand (ActNamespace *, Scope *);
  ActBody *Clone();
private:
  union {
    struct {
      Expr *e;
      const char *msg;
    } t0;
    struct {
      InstType *nu, *old;
    } t1;
    struct {
      int op;
      ActId *id1, *id2;
      const char *msg;
    } t2;
  } u;
  int type;
};

class ActBody_Print : public ActBody {
public:
  ActBody_Print (list_t *_l) {
    l = _l;
  }
  ~ActBody_Print () {
    list_free (l);
  }
  void Expand (ActNamespace *, Scope *);
  ActBody *Clone();
private:
  list_t *l;
};


class ActBody_Namespace : public ActBody {
public:
  ActBody_Namespace (ActNamespace *_ns) {
    ns = _ns;
  }

  void Expand (ActNamespace *, Scope *);
  ActBody *Clone();
  
private:
  ActNamespace *ns;
};


struct act_prs;
struct act_chp;
struct act_refine;
struct act_spec;
struct act_sizing;
struct act_initialize;
struct act_dataflow;

/*
 * Language body
 */
class ActBody_Lang : public ActBody {
 public:

  enum langtype {
    LANG_PRS,
    LANG_CHP,
    LANG_HSE,
    LANG_SPEC,
    LANG_REFINE,
    LANG_SIZE,
    LANG_INIT,
    LANG_DFLOW
  };

  ActBody_Lang (act_prs *p) {
    t = LANG_PRS;
    lang = p;
  }
  ActBody_Lang (act_chp *c, int ishse = 0) {
    if (ishse) {
      t = LANG_HSE;
    }
    else {
      t = LANG_CHP;
    }
    lang = c;
  }

  ActBody_Lang (act_spec *s) {
    t = LANG_SPEC;
    lang = s;
  }
  
  ActBody_Lang (enum langtype _t, void *l) {
    t = _t;
    lang = l;
  }

  ActBody_Lang (act_refine *r) {
    t = LANG_REFINE;
    lang = r;
  }

  ActBody_Lang (act_sizing *s) {
    t = LANG_SIZE;
    lang = s;
  }

  ActBody_Lang (act_initialize *init) {
    t = LANG_INIT;
    lang = init;
  }

  ActBody_Lang (act_dataflow *dflow) {
    t = LANG_DFLOW;
    lang = dflow;
  }
  
  void Expand (ActNamespace *, Scope *);
  void Print (FILE *fp);
  ActBody *Clone();

  void *getlang() { return lang; }
  enum langtype gettype() { return t; }

 private:
  enum langtype t;
  void *lang;
};


/* standard setup/teardown functions for syntactic replication */
void act_syn_loop_setup (ActNamespace *ns, Scope *s,
			    const char *id, Expr *lo, Expr *hi,
			    
			    /* outputs */
			 ValueIdx **vx, int *ilo, int *ihi);

void act_syn_loop_teardown (ActNamespace *ns, Scope *s,
			    const char *id, ValueIdx *vx);


#endif /* __ACT_BODY_H__ */
