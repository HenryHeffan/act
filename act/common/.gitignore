# proprietary files
coi.h
NSOutputInt.h

