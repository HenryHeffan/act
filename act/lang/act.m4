/*************************************************************************
 *
 *  This file is part of the ACT library
 *
 *  Copyright (c) 2011-2019 Rajit Manohar
 *
 *  This program is free software; you can redistribute it and/or
 *  modify it under the terms of the GNU General Public License
 *  as published by the Free Software Foundation; either version 2
 *  of the License, or (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 51 Franklin Street, Fifth Floor,
 *  Boston, MA  02110-1301, USA.
 *
 **************************************************************************
 */

/*------------------------------------------------------------------------
 *
 *  Act top-level pgen file
 *
 *------------------------------------------------------------------------
 */

%type[X] {{ ActTree ActRet }};

/*------------------------------------------------------------------------
 *
 * Grammar
 *
 *------------------------------------------------------------------------
 */
toplevel: [ imports_opens ] [ body ]
{{X:
    OPT_FREE ($1);

    if (!OPT_EMPTY ($2)) {
      ActRet *r;

      r = OPT_VALUE ($2);
      $A(r->type == R_ACT_BODY);
      $0->curns->AppendBody (r->u.body);
      FREE (r);
    }
    OPT_FREE ($2);
    return NULL;
}}
;

imports_opens
: import_open_item imports_opens
| import_open_item
;

body[ActBody *]
: body_item body
{{X:
    if ($1 == NULL) {
      return $2;
    }
    if ($2 == NULL) {
      return $1;
    }
    $1->Append ($2);
    return $1;
}}
| body_item
{{X: return $1; }}
;

ns_body[ActBody *]: ns_body_item ns_body
{{X:
    if ($1 == NULL) {
      return $2;
    }
    if ($2 == NULL) {
      return $1;
    }
    $1->Append ($2);
    return $1;
}}
| ns_body_item
{{X:
    return $1;
}}
;

ns_body_item[ActBody *]
: definition
{{X:
    return NULL;
}}
| namespace_management
{{X:
    ActBody *tmp = new ActBody_Namespace ($1);
    tmp->Append ($1->getBody());
    $1->setBody (NULL);
    tmp->Append (new ActBody_Namespace ($0->curns));
    return tmp;
}}
| namespace_other
{{X: return $1; }}
;

namespace_other[ActBody *]: instance
{{X:
    ActBody_Inst *b;
    $A($1);
    b = dynamic_cast <ActBody_Inst*>($1);
    $A(b);
    if (TypeFactory::isUserType (b->BaseType())) {
      $E("Cannot instantiate user-defined type within a namespace");
    }
    return $1;
}}
| connection
{{X:
    return $1;
}}
| alias
{{X:
    return $1;
}}
;

body_item[ActBody *]
: namespace_management
{{X:
    ActBody *tmp = new ActBody_Namespace ($1);
    tmp->Append ($1->getBody());
    $1->setBody (NULL);
    tmp->Append (new ActBody_Namespace ($0->curns));
    return tmp;
}}
| base_item
{{X: return $1; }}
| definition
{{X: return NULL; }}
;

import_open_item: import_item  | open_item ;

/* namespaces */
include(namespaces.m4)

/* types */
include(types.m4)

/* definitions */
include(defs.m4)

/* languages */
include(lang.m4)

/* expressions */
include(expr.m4)
