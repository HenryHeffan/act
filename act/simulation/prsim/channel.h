/*************************************************************************
 *
 *  Copyright (c) 2005-2019 Rajit Manohar
 *
 *  This program is free software; you can redistribute it and/or
 *  modify it under the terms of the GNU General Public License
 *  as published by the Free Software Foundation; either version 2
 *  of the License, or (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 51 Franklin Street, Fifth Floor,
 *  Boston, MA  02110-1301, USA.
 *
 **************************************************************************
 */
#ifndef __CHANNEL_H__
#define __CHANNEL_H__

#include "prs.h"

#ifdef __cplusplus
extern "C" {
#endif

struct prs_channel;
typedef struct prs_channel PrsChannel;

#define CHINFO(x) ((struct prs_node_extra *)(x)->chinfo)
#define SPACE(x)  ((struct prs_node_extra *)(x)->chinfo)->morespace

// begin CLINT

struct prs_node_extra {
  unsigned int hasChans;
  void *inVector;
  void *morespace;
};

typedef enum {
	CHAN_e1ofN,
	CHAN_eMx1of2,
	CHAN_eMx1of4,
	CHAN_eDx1of2
} ChannelType;

struct prs_channel {
  hash_bucket_t *b;		/* bucket pointer */

	// Channel type
	ChannelType type;

	// Injectfile, dumpfile, and expectfile pointers
	FILE *fpInject, *fpDump, *fpExpect;
	// File names
	char *sInject, *sDump, *sExpect;
	// File line numbers
	int linenumInject, linenumDump, linenumExpect;
	
	// Whether we are doing these things for this channel
	unsigned int isInject:1, isDump:1, isExpect:1;

	// Whether to loop for any of these commands
	unsigned int loopInject:1, loopExpect:1;

	// For every channel that we're watching (dump or expect), it's okay if
	// initially the channel's enable is low but the channel is neutral.
	// Therefore, we keep track of this is the enable's first transition or
	// not...
	unsigned int isFirstEnableTransition:1;

	// Channel name
	char *name;

	// Channel's enable
	PrsNode *enable;

	// Channel's data rails
	PrsNode **dataRails;
	int numDataRails;

	// Size of the channel
	// For e1ofN, this is N.
	// for eMx1of2, eMx1of4, this is M
	int size;

};
// end CLINT

struct Channel {
  // Hashtable of channel names -> channel structs
  struct Hashtable *hChannels;

  // The Reset, required for channel stuff
  PrsNode *reset;
};

struct prs_node_extra *prs_node_extra_init (void);

void channel_checkpoint (struct Channel *C, FILE *fp);
void channel_restore (struct Channel *C, FILE *fp);

  /* Create a new channel */
void create_channel(Prs *, struct Channel *, char *, int , char *);

  /* Drive this channel with values from a file. */
void channel_injectfile(Prs *, struct Channel *, char *, char *, int );

  /* Check that values on this channel match those in a file */
void channel_expectfile(Prs *, struct Channel *, char *, char *, int );

  /* Put this channel's values into a file */ 
void channel_dumpfile(Prs *,  struct Channel *, char *, char *);

  /* Called by prsim whenever an enable switches.  Go through all of
     the channels and call appropriate functions for any channels
     associated with this enable. */
void channel_enableSwitched(Prs *, struct Channel *, PrsNode *);

void channel_resetSwitched(Prs *, struct Channel *);


#ifdef __cplusplus
}
#endif



#endif /* __CHANNEL_H__ */

