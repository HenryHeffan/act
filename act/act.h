/*************************************************************************
 *
 *  This file is part of the ACT library
 *
 *  Copyright (c) 2018-2019 Rajit Manohar
 *
 *  This program is free software; you can redistribute it and/or
 *  modify it under the terms of the GNU General Public License
 *  as published by the Free Software Foundation; either version 2
 *  of the License, or (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 51 Franklin Street, Fifth Floor,
 *  Boston, MA  02110-1301, USA.
 *
 **************************************************************************
 */
#ifndef __ACT_H__
#define __ACT_H__

#include <act/types.h>
#include <act/lang.h>
#include <act/body.h>
#include <act/value.h>
#include <common/config.h>
#include <map>
#include <unordered_set>

/**
 * @file act.h
 *       Contains top-level initialization/management functions for
 *       the ACT library
 */

class ActPass;
class Log;


/*
 *
 *  ACT can model a circuit at different levels of detail. The four
 *  levels are specified below.
 *
 *  The default level is the most detailed model available that can be
 *  provided by the user (PRS)
 *
 */

#define ACT_MODEL_CHP  0		// CHP language
#define ACT_MODEL_HSE  1		// HSE language
#define ACT_MODEL_PRS  2		// PRS language
#define ACT_MODEL_DEVICE 3              // PRS + sizing translated into netlist
#define ACT_MODEL_TOTAL 4

extern const char *act_model_names[];


/**
 *   The main Act class used to read in an ACT file and create basic
 *   data structures
 */
class Act {
 public:
  /**
   * Initialize the ACT library
   *
   *  @param argc is a pointer to argc (command-line processing)
   *  @param argv is a pointer to argv (command-line processing)
   *
   *  @return *argc and *argv are modified to reflect the command-line
   *  options left after ACT has extracted the ones it understands
   */
  static void Init (int *argc, char ***argv);
  static int getOptions (int *argc, char ***argv);
  static void setOptionString (char *str);

  /**
   * Maximum depth of recursion for expanding types
   */
  static int max_recurse_depth;

  /**
   * Maximum number of iterations for general loops
   */
  static int max_loop_iterations;

#define WARNING_FLAG(x,y) \
  static int x ;
#include "warn.def"
  
  /**
   * Command-line arguments if -opt= is used
   */
  static list_t *cmdline_args;

  /**
   * Parser flags 
   */
  static int emit_depend;

  /**
   * Config debugging
   */
  static void config_info (const char *s);
  static void generic_msg (const char *s);


  /**
   * Create an act data structure for the specified input file
   *
   * @param s is the name of the file containing the top-level ACT. If
   * NULL, then the library is initialized without any ACT file being
   * read.
   */
  Act (const char *s = NULL);
  ~Act ();

  /** 
   * Merge in ACT file "s" into current ACT database
   *
   * @param s is the name of an ACT file
   */
  void Merge (const char *s);


  /**
   * Expand types
   */
  void Expand ();


  /**
   * Install string mangling functionality
   *
   * @param s is a string corresponding to the list of characters to
   * be mangled.
   */
  int mangle_set_char (char c);
  
  void mangle (char *s);

  int mangle_active() { return any_mangling; } /**< @return 1 if
						  mangling is active, 
						  0 otherwise */

  /** 
   * mangle string from src to dst.
   * @param src is the source string
   * @param dst is the destination string
   * @param sz is the space available in the destination string
   * @return 0 on success, -1 on error
   */  
  int mangle_string (const char *src, char *dst, int sz);
  
  /** 
   * unmangle string from src to dst.
   * @param src is the source string
   * @param dst is the destination string
   * @param sz is the space available in the destination string
   * @return 0 on success, -1 on error
   */  
  int unmangle_string (const char *src, char *dst, int sz);

  /**
   * Mangle fprintf functionality
   */
  void mfprintf (FILE *fp, const char *s, ...);

  /**
   * Unmangle fprintf functionality
   */
  void ufprintf (FILE *fp, const char *s, ...);

  /**
   * Mangle snprintf functionality
   */
  int msnprintf (char *fp, int sz, const char *s, ...);

  /**
   * Unmangle snprintf functionality
   */
  int usnprintf (char *fp, int sz, const char *s, ...);

  /**
   * Non-standard mangling for user-defined types.
   * @param omit_ns is 1 if you don't want to include the namespace in
   * the string
   */
  void msnprintfproc (char *fp, int sz, UserDef *, int omit_ns = 0);

  /**
   * Non-standard mangling for user-defined types
   * @param omit_ns is 1 if you don't want to include the namespace in
   * the string
   */
  void mfprintfproc (FILE *fp, UserDef *, int omit_ns = 0);

  /* 
     API functions
  */

  /**
   * Find a process given a name
   * @param s is the name of the process
   * @return process pointer, or NULL if not found
   */
  Process *findProcess (const char *s);

  /**
   * Find a process within a namespace
   * @param s is the name of the process
   * @param ns is the ACT namespace
   * @return the process pointer if found, NULL otherwise
   */
  Process *findProcess (ActNamespace *ns, const char *s);

  /**
   * Find a user-defined type
   * @param s is the name of the type
   * @return the UserDef pointer if found, NULL otherwise
   */
  UserDef *findUserdef (const char *s);

  ActNamespace *findNamespace (const char *s);
  ActNamespace *findNamespace (ActNamespace *, const char *);
  ActNamespace *Global() { return gns; }

  /*
    Dump to a file 
  */
  void Print (FILE *fp);

  void pass_register (const char *name, ActPass *p);
  ActPass *pass_find (const char *name);
  void pass_unregister (const char *name);
  const char *pass_name (const char *name);

  /* 
    To mess with types after parsing
  */
  TypeFactory *getTypeFactory () { return tf; }


  /*
    Get modeling level of detail and refinement steps.
  */
  int getLevel ();		// get default level
  int getLevel (Process *p);	// get level for this type
  int getLevel (ActId *id);	// get level for this instance

  int getRefSteps() { return refine_steps; }
  void decRefSteps() { refine_steps--; }
  void incRefSteps() { refine_steps++; }

private:
  TypeFactory *tf;		/* type factory for the file */
  ActNamespace *gns;		/* global namespace */

  char mangle_characters[256];
  int inv_map[256];
  int any_mangling;
  int mangle_langle_idx;  /* index of '<' */
  int mangle_min_idx;     /* index of the min of , . { } */
  int mangle_mode;

  static Log *L;

  struct Hashtable *passes;	// any ActPass-es

  int refine_steps;		// number of refinement steps to
				// process

  int num_type_levels[ACT_MODEL_TOTAL];
  char **type_levels[ACT_MODEL_TOTAL];
  int num_inst_levels[ACT_MODEL_TOTAL];
  char **inst_levels[ACT_MODEL_TOTAL];

  int default_level;  // default level

  static char *_getopt_string;
  
  static int _process_act_arg (const char *argvp, int *tech_specified, char **conf);
};

class ActPass {
protected:
  int _finished;		// has the pass finished execution?
  int _sticky_visited;		// sticky visited flag
  Act *a;			// main act data structure
  list_t *deps;			// ActPass dependencies

  list_t *fwdeps;		// passes that depend on me
  
  const char *name;

  int _update_propagate;
  int _root_dirty;
  Process *_root;
  
  virtual void _actual_update (Process *p);

public:
  ActPass (Act *_a, const char *name, int doroot = 0);
  // Create, initialize, and register pass
  // A pass sets "doroot" to 1 if any update to a process
  // propagates all the way back up to the root of the design.
  // We assume that the list of ports never change.
  
  virtual ~ActPass ();			       // release storage


  int AddDependency (const char *pass); // insert dependency on an
					// actpass

  /*-- re-compute information computed by this pass --*/
  void update (Process *p);

  int rundeps (Process *p = NULL);

  const char *getName();
  
  virtual int run(Process *p = NULL); // run pass on a process; NULL =
				      // top level
  
  int completed()  { return (_finished == 2) ? 1 : 0; }
  int pending()  { return (_finished == 1) ? 1 : 0; }
  void *getMap (Process *p);
  Act *getAct () { return a; }
  ActPass *getPass (const char *name) { return a->pass_find (name); }

  /* -- negative modes are used internally; do not use! -- */
  virtual void run_recursive (Process *p = NULL, int mode = 0);

  void disableUpdate () { _update_propagate = 0; }
  void enableUpdate () { _update_propagate = 1; }


  void mkStickyVisited () { _sticky_visited = 1; }
  void clrStickyVisited() { _sticky_visited = 0; }

  static void refreshAll (Act *a, Process *p = NULL);

private:
  /* -- called before sub-tree -- */
  virtual void *pre_op (Process *p, int mode = 0);
  virtual void *pre_op (Channel *c, int mode = 0);
  virtual void *pre_op (Data *d, int mode = 0);

  /* -- called after sub-tree -- */
  virtual void *local_op (Process *p, int mode = 0);
  virtual void *local_op (Channel *c, int mode = 0);
  virtual void *local_op (Data *d, int mode = 0);

  /* - release storage (if any) -- */
  virtual void free_local (void *);

  int init (); // initialize or re-initialize
  void recursive_op (UserDef *p, int mode = 0);

  void init_map ();
  void free_map ();
  std::map<UserDef *, void *> *pmap;
  std::unordered_set<UserDef *> *visited_flag;
};

struct act_sh_passlib_info {
  char *lib;
  void *lib_ptr;
  int refs;
};

class Technology;
class ActDynamicPass;

struct act_sh_dispatch_table {
  void (*_init) (ActPass *ap);
  void (*_run) (ActPass *ap, Process *p);
  void (*_recursive) (ActPass *ap, Process *p, int mode);
  void *(*_proc) (ActPass *ap, Process *p, int mode);
  void *(*_chan) (ActPass *ap, Channel *c, int mode);
  void *(*_data) (ActPass *ap, Data *d, int mode);
  void (*_free) (ActPass *ap, void *v);
  void (*_done) (ActPass *ap);

  int (*_runcmd) (ActPass *ap, const char *name);
};

class ActDynamicPass : public ActPass {
public:
  ActDynamicPass (Act *_a, const char *name, const char *lib, const char *prefix);
  // load a dynamic pass from a shared object file
  
  ~ActDynamicPass ();		// release storage

  int run (Process *p = NULL);
  void run_recursive (Process *p = NULL, int mode = 0);

  void setParam (const char *name, void *v);
  void setParam (const char *name, int v);
  void setParam (const char *name, double v);
  int getIntParam (const char *name);
  void *getPtrParam (const char *name);
  double getRealParam (const char *name);

  int runcmd (const char *name);

  struct Hashtable *getConfig ();
  Technology *getTech () { return T; }

  bool loaded() { return _load_success; }
  
private:
  virtual void *local_op (Process *p, int mode = 0);
  virtual void *local_op (Channel *c, int mode = 0);
  virtual void *local_op (Data *d, int mode = 0);
  virtual void free_local (void *);

  bool _load_success;
  
  char *_libused;
  act_sh_dispatch_table _d;
  struct Hashtable *_params;
  struct Hashtable *_config_state;
  Technology *T;
  
  /* open shared object libraries */
  static list_t *_sh_libs;
};


/* this should be elsewhere */
Expr *const_expr (long);
Expr *const_expr_bool (int);
Expr *const_expr_real (double);


/*
   Should be called after Act::Init() to add global parameter
   definitions.

   MUST BE CALLED BEFORE an Act object is created!
*/
void act_add_global_pint (const char *name, int val);
void act_add_global_pbool (const char *name, int val);

#endif /* __ACT_H__ */
