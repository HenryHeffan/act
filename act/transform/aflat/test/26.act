import "syn.act";
import "channel.act";
import "globals.act";

a1of2 in[1];
dualrail v[1];
a1of1 go[1];
a1of2 X[1];a1of2 Y[1];a1of2 Z[1];

defproc top (a1of2 in[1];dualrail v[1];a1of1 go[1]){

syn_recv x[1];
x[0] (go[0],v[0],X[0]); 

}

top t;
