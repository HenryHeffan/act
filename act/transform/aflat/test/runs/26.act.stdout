= "Z[0].d[0]" "Z[0].f"
= "Z[0].d[1]" "Z[0].t"
= "Y[0].d[0]" "Y[0].f"
= "Y[0].d[1]" "Y[0].t"
= "X[0].d[0]" "X[0].f"
= "X[0].d[1]" "X[0].t"
= "X[0].a" "x[0].in.a"
= "X[0].f" "x[0].in.f"
= "X[0].t" "x[0].in.t"
= "X[0].d[0]" "x[0].in.d[0]"
= "X[0].d[1]" "x[0].in.d[1]"
= "X[0].t" "X[0].d[1]"
= "X[0].f" "X[0].d[0]"
= "X[0].a" "go[0].a"
= "v[0].d[0]" "v[0].f"
= "v[0].d[1]" "v[0].t"
"t.x[0].go.r"&"t.x[0]._goa"&"t.x[0].in.t"->"t.x[0].v.f"-
"t.x[0].go.r"&"t.x[0]._goa"&"t.x[0].in.f"->"t.x[0].v.t"-
"t.x[0].go.r"&("t.x[0].v.t"&"t.x[0].in.t"|"t.x[0].v.f"&"t.x[0].in.f")->"t.x[0]._goa"-
~"t.x[0].in.t"&~"t.x[0].in.f"&~"t.x[0].go.r"->"t.x[0]._goa"+
"Reset"|"t.x[0]._goa"->"t.x[0].go.a"-
~("Reset"|"t.x[0]._goa")->"t.x[0].go.a"+
= "t.x[0].v.d[0]" "t.x[0].v.f"
= "t.x[0].v.d[1]" "t.x[0].v.t"
= "t.x[0].v.d[0]" "t.x[0].v.f"
= "t.x[0].v.d[1]" "t.x[0].v.t"
= "t.x[0].go.a" "t.x[0].in.a"
= "t.x[0].in.d[0]" "t.x[0].in.f"
= "t.x[0].in.d[1]" "t.x[0].in.t"
= "t.x[0].in.d[0]" "t.x[0].in.f"
= "t.x[0].in.d[1]" "t.x[0].in.t"
= "X[0].a" "t.x[0].in.a"
= "X[0].f" "t.x[0].in.f"
= "X[0].t" "t.x[0].in.t"
= "X[0].d[0]" "t.x[0].in.d[0]"
= "X[0].d[1]" "t.x[0].in.d[1]"
= "t.go[0].a" "t.x[0].go.a"
= "t.go[0].r" "t.x[0].go.r"
= "t.X[0].a" "t.go[0].a"
= "t.v[0].d[0]" "t.v[0].f"
= "t.v[0].d[1]" "t.v[0].t"
= "t.v[0].f" "t.x[0].v.f"
= "t.v[0].t" "t.x[0].v.t"
= "t.v[0].d[0]" "t.x[0].v.d[0]"
= "t.v[0].d[1]" "t.x[0].v.d[1]"
= "t.v[0].d[0]" "t.v[0].f"
= "t.v[0].d[1]" "t.v[0].t"
= "t.in[0].d[0]" "t.in[0].f"
= "t.in[0].d[1]" "t.in[0].t"
= "t.in[0].d[0]" "t.in[0].f"
= "t.in[0].d[1]" "t.in[0].t"
= "in[0].d[0]" "in[0].f"
= "in[0].d[1]" "in[0].t"
